### PR TITLE
[feature](nereids) semi and anti join estimation

### PR DESCRIPTION
**Pull request by englefly**
# Proposed changes
in this pr, we add a new algorithm to estimate semi/anti join row count.
In original alg., we reduce row count from cross join. usually, this is not good.
for example, L left semi join R on L.a=R.a
suppose L is larger than R, and ndv(L.a) < ndv(R.a)
the estimated row count is rowcount(R) * rowcount(L) / ndv(R.a). in most cases, the estimated row count is larger than rowcount(L).

in new alg, we use ndv(R.a)/originalNdv(R.a) to estimate result rowCount. the basic idea is as following:
1. Suppose ndv(R.a) reduced from m to n.
2. Assume that the value space of L.a is the same as R.a if R.a is not filtered.(this assumption is also hold in original alg.)
regard `L left join R` as a filter applied on L, that is, if L.a is in R.a, then this tuple stays in result.
R.a shrinks to m/n, so L.a also shrinks to m/n

Issue Number: close #xxx

## Problem summary

Describe your changes.

## Checklist(Required)

* [ ] Does it affect the original behavior
* [ ] Has unit tests been added
* [ ] Has document been added or modified
* [ ] Does it need to update dependencies
* [ ] Is this PR support rollback (If NO, please explain WHY)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/nereids/stats/JoinEstimation.java [hunk 1]
```diff
@@ -60,11 +63,84 @@ private static double estimateJoinConditionSel(Statistics crossJoinStats, Expres
...
+            rowCount = leftStats.getRowCount() * buildColStats.ndv / buildColStats.originalNdv;
+        } else {
+            //right semi or anti
+            rowCount = rightStats.getRowCount() * probColStats.ndv / probColStats.originalNdv;
+        }
+        return rowCount;
+    }
+    private static Statistics estimateSemiOrAnti(Statistics leftStats, Statistics rightStats, Join join) {
+        // primaryConjunct is the most effective conjunct.
+        Expression primaryConjunct;
```
Comment (Kikyou1997) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/stats/JoinEstimation.java:103]: This variable never gets used

File: fe/fe-core/src/main/java/org/apache/doris/nereids/stats/JoinEstimation.java [hunk 2]
```diff
@@ -73,16 +149,15 @@ public static Statistics estimate(Statistics leftStats, Statistics rightStats, J
...
             innerJoinStats = crossJoinStats;
         }
-        // if (!join.getOtherJoinConjuncts().isEmpty()) {
-        //     FilterEstimation filterEstimation = new FilterEstimation();
-        //     innerJoinStats = filterEstimation.estimate(
-        //             ExpressionUtils.and(join.getOtherJoinConjuncts()), innerJoinStats);
-        // }
+        if (!join.getOtherJoinConjuncts().isEmpty()) {
+            FilterEstimation filterEstimation = new FilterEstimation();
+            innerJoinStats = filterEstimation.estimate(
```
Comment (Kikyou1997) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/stats/JoinEstimation.java:158]: Move this logic to `estimateInnerJoin` plz

File: fe/fe-core/src/test/java/org/apache/doris/nereids/util/HyperGraphBuilder.java [hunk 3]
```diff
@@ -190,7 +190,7 @@ private void injectRowcount(Group group) {
             int count = rowCounts.get(Integer.parseInt(scanPlan.getTable().getName()));
             for (Slot slot : scanPlan.getOutput()) {
                 slotIdToColumnStats.put(slot,
-                        new ColumnStatistic(count, count, 0, 0, 0, 0,
+                        new ColumnStatistic(count, count, 0, 0,0, 0, 0,
```
Comment (Kikyou1997) [inline fe/fe-core/src/test/java/org/apache/doris/nereids/util/HyperGraphBuilder.java:193]: Fix code style

File: fe/fe-core/src/main/java/org/apache/doris/nereids/stats/JoinEstimation.java [hunk 4]
```diff
@@ -60,11 +63,84 @@ private static double estimateJoinConditionSel(Statistics crossJoinStats, Expres
...
+        double rowCount;
+        if (join.getJoinType().isLeftSemiOrAntiJoin()) {
+            rowCount = leftStats.getRowCount() * buildColStats.ndv / buildColStats.originalNdv;
+        } else {
+            //right semi or anti
+            rowCount = rightStats.getRowCount() * probColStats.ndv / probColStats.originalNdv;
+        }
+        return rowCount;
+    }
+    private static Statistics estimateSemiOrAnti(Statistics leftStats, Statistics rightStats, Join join) {
```
Comment (Kikyou1997) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/stats/JoinEstimation.java:101]: Fix code style

**Comment by Kikyou1997** [comment]
semms originalNDV only set when loading cache,  would you like to add some comment to explain how this MORE PRECISE estimation algorithm works?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/stats/JoinEstimation.java (hunk 2 above)
Comment (englefly) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/stats/JoinEstimation.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/stats/JoinEstimation.java (hunk 1 above)
Comment (englefly) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/stats/JoinEstimation.java]: removed

**Comment by englefly** [comment]
> 

explained  in 'proposed changes'

**Comment by englefly** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 35.79 seconds
 stream load tsv:          472 seconds loaded 74807831229 Bytes, about 151 MB/s
 stream load json:         22 seconds loaded 2358488459 Bytes, about 102 MB/s
 stream load orc:          74 seconds loaded 1101869774 Bytes, about 14 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230328010622_clickbench_pr_121529.html

File: fe/fe-core/src/main/java/org/apache/doris/nereids/stats/JoinEstimation.java [hunk 5]
```diff
@@ -52,37 +55,105 @@ private static Statistics estimateInnerJoin(Statistics crossJoinStats, List<Expr
...
+        return innerJoinStats;
     }
 
     private static double estimateJoinConditionSel(Statistics crossJoinStats, Expression joinCond) {
         Statistics statistics = new FilterEstimation().estimate(joinCond, crossJoinStats);
         return statistics.getRowCount() / crossJoinStats.getRowCount();
     }
 
+    private static double adjustSemiOrAntiByOtherJoinConditions(Join join) {
+        final double non_equal_ratio = 0.5;
```
Comment (Kikyou1997) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/stats/JoinEstimation.java:74]: We use camel case naming style

Comment (Kikyou1997) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/stats/JoinEstimation.java:74]: Besides, similar variable has been defined in FilterEstimation with public static modifier

File: fe/fe-core/src/main/java/org/apache/doris/nereids/stats/JoinEstimation.java [hunk 6]
```diff
@@ -52,37 +55,105 @@ private static Statistics estimateInnerJoin(Statistics crossJoinStats, List<Expr
...
+            rowCount = leftStats.getRowCount() * buildColStats.ndv / buildColStats.originalNdv;
+        } else {
+            //right semi or anti
+            rowCount = rightStats.getRowCount() * probColStats.ndv / probColStats.originalNdv;
+        }
+        return rowCount;
+    }
+
+    private static Statistics estimateSemiOrAnti(Statistics leftStats, Statistics rightStats, Join join) {
+        // primaryConjunct is the most effective conjunct.
```
Comment (Kikyou1997) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/stats/JoinEstimation.java:110]: Remove this redundant comment plz

File: fe/fe-core/src/main/java/org/apache/doris/nereids/stats/JoinEstimation.java [hunk 7]
```diff
@@ -52,37 +55,105 @@ private static Statistics estimateInnerJoin(Statistics crossJoinStats, List<Expr
...
+        } else {
+            buildColStats = rightStats.findColumnStatistics(eqRight);
+        }
+        if (probColStats == null || buildColStats == null) {
+            return Double.POSITIVE_INFINITY;
+        }
+
+        double rowCount;
+        if (join.getJoinType().isLeftSemiOrAntiJoin()) {
+            rowCount = leftStats.getRowCount() * buildColStats.ndv / buildColStats.originalNdv;
```
Comment (Kikyou1997) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/stats/JoinEstimation.java:101]: Use StatsMathUtil to do `divide` to avoid ArithmeticException

File: fe/fe-core/src/main/java/org/apache/doris/nereids/stats/JoinEstimation.java [hunk 8]
```diff
@@ -52,37 +55,105 @@ private static Statistics estimateInnerJoin(Statistics crossJoinStats, List<Expr
...
+    private static Statistics estimateSemiOrAnti(Statistics leftStats, Statistics rightStats, Join join) {
+        // primaryConjunct is the most effective conjunct.
+        double rowCount = Double.POSITIVE_INFINITY;
+        for (Expression conjunct : join.getHashJoinConjuncts()) {
+            double eqRowCount = estimateSemiOrAntiRowCountByEqual(leftStats, rightStats, join, (EqualTo) conjunct);
+            if (rowCount > eqRowCount) {
+                rowCount = eqRowCount;
+            }
+        }
+        if (rowCount == Double.POSITIVE_INFINITY) {
```
Comment (Kikyou1997) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/stats/JoinEstimation.java:118]: Use Double.isInfinite() may be better

File: fe/fe-core/src/main/java/org/apache/doris/nereids/stats/JoinEstimation.java [hunk 9]
```diff
@@ -52,37 +55,105 @@ private static Statistics estimateInnerJoin(Statistics crossJoinStats, List<Expr
...
+            //fall back to original alg.
+            return null;
+        }
+        rowCount = rowCount * adjustSemiOrAntiByOtherJoinConditions(join);
+
+        StatisticsBuilder builder;
+        if (join.getJoinType().isLeftSemiOrAntiJoin()) {
+            leftStats.fix(rowCount, leftStats.getRowCount());
```
Comment (Kikyou1997) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/stats/JoinEstimation.java:126]: `fix` method here actually changed the stats of child which should be avoided as much as possible.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/stats/JoinEstimation.java [hunk 10]
```diff
@@ -52,37 +55,105 @@ private static Statistics estimateInnerJoin(Statistics crossJoinStats, List<Expr
...
+            builder = new StatisticsBuilder(leftStats);
+            builder.setRowCount(rowCount);
+        } else {
+            //right semi or anti
+            rightStats.fix(rowCount, rightStats.getRowCount());
```
Comment (Kikyou1997) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/stats/JoinEstimation.java:131]: ditto

**Review by Kikyou1997 — CHANGES_REQUESTED** [review]
Need more fix

File: fe/fe-core/src/main/java/org/apache/doris/nereids/stats/JoinEstimation.java [hunk 11]
```diff
@@ -52,37 +55,105 @@ private static Statistics estimateInnerJoin(Statistics crossJoinStats, List<Expr
         for (int i = 0; i < sortedJoinConditions.size(); i++) {
             sel *= Math.pow(sortedJoinConditions.get(i).second, 1 / Math.pow(2, i));
         }
-        return crossJoinStats.updateRowCountOnly(crossJoinStats.getRowCount() * sel);
+        Statistics innerJoinStats = crossJoinStats.updateRowCountOnly(crossJoinStats.getRowCount() * sel);
+
+        if (!join.getOtherJoinConjuncts().isEmpty()) {
```
Comment (Kikyou1997) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/stats/JoinEstimation.java:60]: Is this change necessary?

File: fe/fe-core/src/main/java/org/apache/doris/statistics/ColumnStatisticBuilder.java [hunk 12]
```diff
@@ -35,6 +35,8 @@ public class ColumnStatisticBuilder {
 
     private Histogram histogram;
 
+    private double originalNdv;
```
Comment (Kikyou1997) [inline fe/fe-core/src/main/java/org/apache/doris/statistics/ColumnStatisticBuilder.java:38]: Add some comment for this member

File: fe/fe-core/src/main/java/org/apache/doris/nereids/stats/JoinEstimation.java [hunk 13]
```diff
@@ -52,37 +55,105 @@ private static Statistics estimateInnerJoin(Statistics crossJoinStats, List<Expr
...
+        }
+        return builder.build();
+
+    }
+
     /**
      * estimate join
      */
     public static Statistics estimate(Statistics leftStats, Statistics rightStats, Join join) {
         JoinType joinType = join.getJoinType();
```
Comment (Kikyou1997) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/stats/JoinEstimation.java:143]: Codes of this method looks mess, better to refactor it

**Review by Kikyou1997 — CHANGES_REQUESTED** [review]
Besides, I run this codes on my local machine found qError for q21 Increased by more than 1000 times compared to before and caused the avg qerror for tpch-h 1G queries increased by more thatn 3 tiems. Please figure it out and solve this

File: fe/fe-core/src/main/java/org/apache/doris/nereids/stats/JoinEstimation.java (hunk 5 above)
Comment (englefly) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/stats/JoinEstimation.java:74]: Ok,  non_equal_ratio  is replaced by FilterEstimation.DEFAULT_INEQUALITY_COEFFICIENT.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/stats/JoinEstimation.java (hunk 6 above)
Comment (englefly) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/stats/JoinEstimation.java:110]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/stats/JoinEstimation.java (hunk 8 above)
Comment (englefly) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/stats/JoinEstimation.java:118]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/stats/JoinEstimation.java (hunk 7 above)
Comment (englefly) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/stats/JoinEstimation.java:101]: done

**Comment by englefly** [comment]
run buildall

**Review by Kikyou1997 — DISMISSED** [review]

File: fe/fe-core/src/main/java/org/apache/doris/nereids/stats/JoinEstimation.java (hunk 11 above)
Comment (englefly) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/stats/JoinEstimation.java:65]: yes. OtherJoinCondition should be counted in separately.
JoinCondition and otherJoinCondition should not disturb each other

File: fe/fe-core/src/main/java/org/apache/doris/nereids/stats/JoinEstimation.java (hunk 13 above)
Comment (englefly) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/stats/JoinEstimation.java:153]: done

**Comment by englefly** [comment]
run buildall

**Comment by englefly** [comment]
> Besides, I run this codes on my local machine found qError for q21 Increased by more than 1000 times compared to before and caused the avg qerror for tpch-h 1G queries increased by more thatn 3 tiems. Please figure it out and solve this

qError is meaningful only if two plans have the same join order.
In this pr, q21 plan join order changes a lot.

**Comment by englefly** [comment]
run buildall

**Comment by englefly** [comment]
run buildp0

**Comment by englefly** [comment]
run buildall

**Review by Kikyou1997 — APPROVED** [review]

File: fe/fe-core/src/main/java/org/apache/doris/statistics/ColumnStatisticBuilder.java (hunk 12 above)
Comment (englefly) [reply fe/fe-core/src/main/java/org/apache/doris/statistics/ColumnStatisticBuilder.java:38]: done

**Review by starocean999 — APPROVED** [review]
